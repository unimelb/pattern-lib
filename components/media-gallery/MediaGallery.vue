<template>
  <div>
    <div
      v-if="!showThumbnails"
      :class="classes"
      role="region"
      aria-label="Media gallery">
      <div :class="containerClasses">
        <div
          v-if="useOverlay"
          class="media-gallery__header">
          <div class="media-gallery__count">
            {{ selectedIndex + 1 }} / {{ items.length }}
          </div>
          <div
            class="media-gallery__close"
            tabindex="0"
            title="Close (Esc)"
            @click="openStateToggle"
            @keydown.enter="openStateToggle"
            @keydown.space="openStateToggle">
            <SvgIcon
              aria-hidden="true"
              name="close"
              width="26"
              height="26" />
          </div>
        </div>
        <div :class="sliderClasses">
          <div
            class="media-gallery__button media-gallery__button--prev"
            role="button"
            tabindex="0"
            title="Previous (arrow left)"
            @click="move('prev')"
            @keydown.enter="move('prev')"
            @keydown.space="move('prev')">
            <SvgIcon
              aria-hidden="true"
              name="chevron-left"
              width="30"
              height="30" />
          </div>
          <div
            v-if="items.length"
            class="media-gallery__slider-container">
            <slider
              v-if="!useOverlay || openState"
              ref="slider"
              :options="options"
              @slide="slide">
              <slideritem
                v-for="(item, index) in items"
                :key="index"
                class="media-gallery__item">
                <div
                  v-if="item.type === 'image'"
                  :aria-label="item.altText"
                  :style="{ backgroundImage: `url(${item.src})` }"
                  class="media-gallery__image" />
                <VideoEmbed
                  v-if="item.type === 'video'"
                  :src="item.src"
                  class="media-gallery__embed" />
              </slideritem>
            </slider>
          </div>
          <div
            class="media-gallery__button media-gallery__button--next"
            tabindex="0"
            role="button"
            title="Next (arrow right)"
            @click="move('next')"
            @keydown.enter="move('next')"
            @keydown.space="move('next')">
            <SvgIcon
              aria-hidden="true"
              name="chevron-right"
              width="30"
              height="30" />
          </div>
        </div>

        <div
          v-if="items.length"
          ref="thumbnailContainer"
          :class="thumbClasses">
          <div
            v-for="(item, index) in items"
            :key="index"
            :class="{ active: index === selectedIndex }"
            :aria-label="item.title + '. ' + item.description"
            class="media-gallery__thumb"
            tabindex="0"
            role="button"
            @click="open(index)">
            <img
              v-if="item.type === 'image'"
              :src="item.src"
              :alt="item.altText"
              aria-hidden="true"
              class="media-gallery__thumb-image">
            <div
              v-if="item.type === 'video'"
              aria-hidden="true"
              class="media-gallery__thumb-embed">
              <VideoEmbed
                :src="item.src"
                class="media-gallery__thumb-video" />
            </div>
          </div>
        </div>

        <div
          v-if="items.length"
          class="media-gallery__footer">
          <div
            v-if="!useOverlay"
            class="media-gallery__count media-gallery__count--footer">
            {{ selectedIndex + 1 }} / {{ items.length }}
          </div>
          <div
            :id="'caption' + selectedIndex"
            class="media-gallery__title">
            {{ selectedItem.title }}
          </div>
          <div class="media-gallery__description">
            {{ selectedItem.description }}
          </div>
        </div>
      </div>
    </div>
    <ThumbnailGallery
      v-if="useOverlay || showThumbnails"
      :use-overlay="useOverlay"
      :items="items"
      :callback="openThumb"
      :display-caption="displayCaption"
      :columns="columns" />
  </div>
</template>

<script>
import { slider, slideritem } from 'vue-concise-slider';
import debounce from 'lodash.debounce';
import VideoEmbed from '../embed/VideoEmbed.vue';
import SvgIcon from '../icons/SvgIcon.vue';
import ThumbnailGallery from './ThumbnailGallery.vue';

import { KEYCODE_ESC, KEYCODE_LEFT, KEYCODE_RIGHT } from '../../constants/keycodes';
import { TIMER_100 } from '../../constants/timers';

export default {
  components: {
    VideoEmbed,
    slider,
    slideritem,
    SvgIcon,
    ThumbnailGallery,
  },
  props: {
    columns: {
      type: String,
      default: '4',
    },
    items: {
      type: Array,
      default: () => [],
    },
    useOverlay: {
      type: Boolean,
      default: false,
    },
    displayCaption: {
      type: Boolean,
      default: false,
    },
    showThumbnails: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      selectedItem: this.items ? this.items[0] : {},
      selectedIndex: 0,
      openState: false,
      options: {
        currentPage: 0,
        pagination: false,
        thresholdDistance: 100, // Sliding distance threshold
        thresholdTime: 300, // Sliding time threshold decision
        grabCursor: true, // Scratch style
        speed: 500, // Sliding speed
        timingFunction: 'ease', // Sliding mode
        itemAnimation: true,
      },
      isInViewport: false,
    };
  },
  computed: {
    classes() {
      return {
        'media-gallery': true,
        'media-gallery--overlay': this.useOverlay,
        'media-gallery--hide': this.useOverlay && !this.openState,
        'media-gallery--show': this.openState,
      };
    },
    sliderClasses() {
      return {
        'media-gallery__slider': true,
        'media-gallery__slider--overlay': this.useOverlay,
      };
    },
    containerClasses() {
      return {
        'media-gallery__container': true,
        'media-gallery__container--overlay': this.useOverlay,
      };
    },
    thumbClasses() {
      return {
        'media-gallery__thumbnails': true,
        'media-gallery__thumbnails--overlay': this.useOverlay,
      };
    },
  },
  mounted() {
    window.addEventListener('keyup', this.keyBoardActions);

    this.debouncedMediaGalleryScrollEvent = debounce(this.checkInViewport, TIMER_100);
    window.addEventListener('scroll', this.debouncedMediaGalleryScrollEvent);
  },
  beforeDestroy() {
    window.removeEventListener('keyup', this.keyBoardActions);
    window.removeEventListener('scroll', this.debouncedMediaGalleryScrollEvent);
  },
  methods: {
    open(index) {
      this.setSelectedItem(this.items[index]);

      this.setSelectedIndex(index);

      if (this.$refs.slider) {
        this.slideTo(index);

        this.scrollToView(index);

        if (this.$refs.thumbnailContainer.childNodes[index].querySelector('iframe')) {
          this.stopVideo();
        }
      }
    },
    openThumb(index) {
      this.options.currentPage = index;
      this.openStateToggle();
    },
    openStateToggle() {
      this.openState = !this.openState;

      this.toggleNoScroll();
    },
    move(direction) {
      const { items } = this;
      const currentIndex = this.getSelectedIndex();
      const nextIndex = this.getNextIndex(direction);
      const nextItem = items[nextIndex];

      this.setSelectedItem(nextItem);

      this.setSelectedIndex(nextIndex);

      if (this.$refs.slider) {
        this.slideTo(nextIndex);

        this.scrollToView(nextIndex);

        if (this.$refs.thumbnailContainer.childNodes[currentIndex].querySelector('iframe')) {
          this.stopVideo();
        }
      }
    },
    getNextIndex(direction) {
      const itemsLength = this.items.length;
      const currentIndex = this.getSelectedIndex();
      const directions = {
        prev: (currentIndex + itemsLength - 1) % itemsLength,
        next: (currentIndex + 1) % itemsLength,
      };

      const nextIndex = directions[direction];

      return nextIndex;
    },
    setSelectedItem(index) {
      this.selectedItem = index;
    },
    setSelectedIndex(index) {
      this.selectedIndex = index;
    },
    getSelectedIndex() {
      return this.selectedIndex;
    },
    // Scroll active thumbnail into view
    scrollToView(nextIndex) {
      const elementToScrollTo = this.$refs.thumbnailContainer.childNodes[nextIndex];

      // Only scroll element if component is in viewport.
      if (this.isInViewport) {
        elementToScrollTo.scrollIntoView({ behavior: 'smooth', block: 'nearest', inline: 'center' });
      }
    },
    stopVideo() {
      const iframe = document.querySelectorAll('iframe');
      if (iframe.length) {
        iframe.forEach((element) => {
          const iframeSrc = element.src;
          element.src = iframeSrc;
        });
      }
    },
    slide(slide) {
      const { currentPage } = slide;

      this.scrollToView(currentPage);

      this.setSelectedItem(this.items[currentPage]);
      this.setSelectedIndex(currentPage);
    },
    slideTo(index) {
      this.$refs.slider.$emit('slideTo', index);
    },
    keyBoardActions(e) {
      // Only if component is in viewport.
      if (this.isInViewport) {
        if (e.keyCode === KEYCODE_LEFT) {
          this.move('prev');
        }
        if (e.keyCode === KEYCODE_RIGHT) {
          this.move('next');
        }
        if (e.keyCode === KEYCODE_ESC) {
          this.openState = false;
        }
      }
    },
    toggleNoScroll() {
      document.documentElement.classList.toggle('no-body-scroll');
    },
    checkInViewport() {
      const elementToCheck = this.$el;
      const isInViewport = this.isAnyPartOfElementInViewport(elementToCheck);

      this.isInViewport = isInViewport;
    },
    isAnyPartOfElementInViewport(el) {
      const elementRect = el.getBoundingClientRect();

      const windowHeight = (window.innerHeight || document.documentElement.clientHeight);
      const windowWidth = (window.innerWidth || document.documentElement.clientWidth);

      const offsetTop = 0;
      const offsetLeft = 0;

      const vertInView = (elementRect.top <= windowHeight) && ((elementRect.top + elementRect.height) >= offsetTop);
      const horInView = (elementRect.left <= windowWidth) && ((elementRect.left + elementRect.width) >= offsetLeft);

      return (vertInView && horInView);
    },
  },
};
</script>
