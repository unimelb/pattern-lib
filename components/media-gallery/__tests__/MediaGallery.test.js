import {
  shallow,
  mount,
} from 'vue-test-utils';
import { axe, toHaveNoViolations } from 'jest-axe';
import sinon from 'sinon';
import MediaGallery from '../MediaGallery.vue';

expect.extend(toHaveNoViolations);

const itemsMock = [{
  src: 'test',
  title: 'test1',
  description: 'test1',
  type: 'image',
  altText: 'alt text 1',
},
{
  src: 'test2',
  title: 'test2',
  description: 'test2',
  type: 'video',
  altText: 'alt text 2',
},
];

describe('MediaGallery Common', () => {
  it('should match snapshot', () => {
    const result = shallow(MediaGallery).element;
    expect(result).toMatchSnapshot();
  });

  it('should have default props and correct types items, useOverlay, columns, displayCaption', () => {
    const wrapper = shallow(MediaGallery);
    const {
      items, useOverlay, columns, displayCaption,
    } = wrapper.vm.$options.props;

    expect(columns.type).toBe(String);
    expect(wrapper.props().columns).toBe('4');
    expect(items.type).toBe(Array);
    expect(useOverlay.type).toBe(Boolean);
    expect(displayCaption.type).toBe(Boolean);
    expect(wrapper.props().items).toEqual([]);
  });

  it('should accept cols prop with correct type', () => {
    const columns = '3';
    const wrapper = shallow(MediaGallery, {
      propsData: {
        columns,
      },
    });

    expect(typeof wrapper.props().columns).toBe('string');
    expect(wrapper.props().columns).toBe(columns);
  });

  it('should render media image in slider from prop', () => {
    const wrapper = shallow(MediaGallery, {
      propsData: {
        items: itemsMock,
      },
    });
    expect(wrapper.props().items.length).toBe(2);
    expect(wrapper.find('img').attributes().src).toBe('test');
  });

  it('should render media video in slider from prop', () => {
    const wrapper = mount(MediaGallery, {
      propsData: {
        items: itemsMock,
      },
    });
    expect(wrapper.props().items.length).toBe(2);
    expect(wrapper.find('iframe').exists()).toBe(true);
  });

  it('should set active to first media as default', () => {
    const wrapper = mount(MediaGallery, {
      propsData: {
        items: itemsMock,
      },
    });
    expect(wrapper.vm.selectedIndex).toBe(0);
  });

  it('should have click event on arrows', () => {
    const move = sinon.stub();
    const wrapper = shallow(MediaGallery, {
      propsData: {
        items: itemsMock,
      },
    });

    wrapper.setMethods({
      move,
    });
    wrapper.find('.media-gallery__button').trigger('click');
    expect(move.called).toBe(true);
  });

  it('should have click event on thumb', () => {
    const open = sinon.stub();
    const wrapper = shallow(MediaGallery, {
      propsData: {
        items: itemsMock,
      },
    });

    wrapper.setMethods({
      open,
    });
    wrapper.find('.media-gallery__thumb-image').trigger('click');
    expect(open.called).toBe(true);
  });

  it('should change media on arrow click', () => {
    const wrapper = shallow(MediaGallery, {
      propsData: {
        items: itemsMock,
      },
    });

    Element.prototype.scrollIntoView = jest.fn();

    wrapper.find('.media-gallery__button').trigger('click');
    expect(wrapper.vm.selectedIndex).toBe(1);
  });

  it('should have click event on arrows', () => {
    const move = sinon.stub();
    const wrapper = shallow(MediaGallery, {
      propsData: {
        items: itemsMock,
      },
    });

    wrapper.setMethods({
      move,
    });
    wrapper.find('.media-gallery__button').trigger('keydown.enter');
    expect(move.called).toBe(true);
  });

  it('should render current image title/description', () => {
    const wrapper = mount(MediaGallery, {
      propsData: {
        items: itemsMock,
      },
    });

    Element.prototype.scrollIntoView = jest.fn();

    expect(wrapper.vm.selectedIndex).toBe(0);
    expect(wrapper.find('.media-gallery__title').text()).toBe('test1');
    expect(wrapper.find('.media-gallery__description').text()).toBe('test1');
    wrapper.find('.media-gallery__button').trigger('click');
    expect(wrapper.find('.media-gallery__title').text()).toBe('test2');
    expect(wrapper.find('.media-gallery__description').text()).toBe('test2');
  });
});

describe('MediaGallery InPage', () => {
  it('should render current media number', () => {
    const wrapper = mount(MediaGallery, {
      propsData: {
        items: itemsMock,
      },
    });

    Element.prototype.scrollIntoView = jest.fn();

    expect(wrapper.find('.media-gallery__count').text()).toBe('1 / 2');
    wrapper.find('.media-gallery__button').trigger('click');
    expect(wrapper.find('.media-gallery__count').text()).toBe('2 / 2');
  });

  it('should change image on thumb click', () => {
    const wrapper = shallow(MediaGallery, {
      propsData: {
        items: itemsMock,
      },
    });
    wrapper.findAll('.media-gallery__thumb').at(1).trigger('click');
    expect(wrapper.findAll('.media-gallery__thumb').at(1).classes()).toContain('active');
    expect(wrapper.vm.selectedIndex).toBe(1);
  });

  it('should render embed if image type', () => {
    const wrapper = mount(MediaGallery, {
      propsData: {
        items: itemsMock,
      },
    });
    wrapper.find('.media-gallery__button').trigger('click');
    wrapper.find('.media-gallery__button').trigger('click');
    expect(wrapper.find('.media-gallery__thumb-image').exists()).toBe(true);
  });

  it('should render embed if video type', () => {
    const wrapper = mount(MediaGallery, {
      propsData: {
        items: itemsMock,
      },
    });
    wrapper.find('.media-gallery__button').trigger('click');
    expect(wrapper.find('.embed').exists()).toBe(true);
  });
});

describe('MediaGallery overlay', () => {
  it('should have overlay classes', () => {
    const wrapper = mount(MediaGallery, {
      propsData: {
        items: itemsMock,
        useOverlay: true,
      },
    });

    expect(wrapper.find('.media-gallery').classes()).toContain('media-gallery--hide');
    expect(wrapper.find('.media-gallery__count').exists()).toBe(true);
    expect(wrapper.find('.media-gallery__close').exists()).toBe(true);
    expect(wrapper.find('.media-gallery__slider').classes()).toContain('media-gallery__slider--overlay');
    expect(wrapper.find('.media-gallery__container--overlay').exists()).toBe(true);
  });

  it('should render current media number', () => {
    const wrapper = shallow(MediaGallery, {
      propsData: {
        items: itemsMock,
        useOverlay: true,
      },
    });

    expect(wrapper.find('.media-gallery__count').text()).toBe('1 / 2');
    wrapper.find('.media-gallery__button').trigger('click');
    expect(wrapper.find('.media-gallery__count').text()).toBe('2 / 2');
  });

  it('should have thumbnails component', () => {
    const wrapper = mount(MediaGallery, {
      propsData: {
        items: itemsMock,
        useOverlay: true,
      },
    });

    expect(wrapper.find('.thumbnails').exists()).toBe(true);
  });

  it('should have img thumbnails open callback', () => {
    const openThumb = sinon.stub();
    const wrapper = mount(MediaGallery, {
      propsData: {
        items: itemsMock,
        useOverlay: true,
      },
    });

    wrapper.setMethods({
      openThumb,
    });
    wrapper.find('.thumbnails__item div').trigger('click');
    expect(openThumb.called).toBe(true);
  });

  it('should open slider pop over on thumb click', () => {
    const wrapper = mount(MediaGallery, {
      propsData: {
        items: itemsMock,
        useOverlay: true,
      },
    });

    wrapper.find('.thumbnails__item div').trigger('click');
    expect(wrapper.vm.openState).toBe(true);
    expect(wrapper.find('.media-gallery').classes()).toContain('media-gallery--show');
  });

  it('Component throws no accessibility violations', (done) => {
    const html = shallow(MediaGallery, {
      propsData: {
        items: itemsMock,
      },
    }).html();
    // pass anything that outputs html to axe
    return axe(html).then((response) => {
      expect(response).toHaveNoViolations();
      done();
    });
  });
});
