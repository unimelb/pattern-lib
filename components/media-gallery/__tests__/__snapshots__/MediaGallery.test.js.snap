// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaGallery Common should match snapshot 1`] = `
<div>
  <div
    aria-label="Media gallery"
    class="media-gallery"
    role="region"
  >
    <div
      class="media-gallery__container"
    >
      <!---->
       
      <div
        class="media-gallery__slider"
      >
        <div
          class="media-gallery__button media-gallery__button--prev"
          role="button"
          tabindex="0"
          title="Previous (arrow left)"
        >
          <!---->
        </div>
         
        <!---->
         
        <div
          class="media-gallery__button media-gallery__button--next"
          role="button"
          tabindex="0"
          title="Next (arrow right)"
        >
          <!---->
        </div>
      </div>
       
      <!---->
       
      <!---->
    </div>
  </div>
   
  <!---->
</div>
`;
